### PR TITLE
Lists: Undo Exceptions Mark V

### LINKED ISSUE #467: UndoManager: Exceptions
### Description:
The UndoManager ends up throwing an exception whenever the document being edited contains a Checklist.

1. New Document
2. Insert a new Checklist
3. Add a few lines
4. Hit CTRL + Z few times

As a result, you'll get an exception in the Console:

```
2020-03-12 12:31:48.347792-0300 Simplenote[54444:871909] [General] *** -[Swift.__StringStorage substringWithRange:]: Range {9, 11} out of bounds; string length 16
2020-03-12 12:31:48.350566-0300 Simplenote[54444:871909] [General] (
	0   CoreFoundation                      0x00007fff2e3e98ab __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff646a3805 objc_exception_throw + 48
	2   Foundation                          0x00007fff309da61d -[NSString substringWithRange:] + 223
	3   AppKit                              0x00007fff2b78397d -[NSTextStorage(NSUndo) _undoRedoAttributedSubstringFromRange:] + 145
	4   AppKit                              0x00007fff2bd141d7 -[NSUndoTyping undoRedo:] + 142
	5   Foundation                          0x00007fff30b08a99 -[_NSUndoStack popAndInvoke] + 187
```

### Details:
1. Checklists are internally represented by means of **NSTextAttachments** embedded within the main **NSTextStorage**
2. Documents stored in Simperium encode checklists as `- [ ]` or `- [x]`
3. Inserting a new checklist results in **SPTextView** injecting a `- [ ]` string, and then running a RegEx that turns such substring into a **NSTextAttachment** (for Image purposes)

What's the problem:
1. The Regex / Replacement happen at the **NSTextStorage** level, and thus, the Undo Manager is unaware of the latest changes
2. Hitting CTRL + Z attempts to remove the `- [ ]` marker (when instead we've got an attachment of length 1)

Even if we patched up the NSTextAttachment replacement, hitting CTRL + Z would end up stripping the Checklist Image, which is **Definitely** not what we want.

### Tasks:
- [x] New **NSAttributedString** to **Markdown** Converter (saving the document)
- [x] Hitting **newline** should also insert a **NSTextAttachment** _directly_
- [x] Inserting a Checklist should insert a **NSTextAttachment** _directly_
- [x] New **NSTextView.displayNoteWithContents** API: Process Lists + Reset UndoManager . This prevents List Processing from being undoable.
- [x] Typing `- [ ]` or `- [x]` should still cause the app to inject a checklist

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details:
As of this PR, whenever a user types `- [ ]` we'll perform a realtime replacement with the appropriate **TextAttachment** counterpart.

Pressing **CMD + Z** is expected to yield `- [ ` (that is, the exact string you had prior to the bullet showing up), and pressing **Shift + CMD + Z** should redo the OP.

End of the line PR, Undo Exceptions should be gone... or at least waaaay harder to trigger.

@aerych @diegoreymendez Invoking the both of you to ... try to break things up, whenever you have a few.

**Thanks in advance!!!!** 🔥🔥🔥

Closes #467

---

### Scenario A: Typing List Items
1. Add a new note
2. Type `- [ ]` (character by character please!)

- [x] Verify a List Item immediately shows up
- [x] Verify that pressing **CMD + Z** results in `- [ ` onscreen, and the cursor is at the expected location
- [x] Verify in a web browser that (Undo) change is effectively propagated thru the backend
- [x] Verify that pressing **Shift + CMD + Z** performs a REDO!

### Scenario B: Smoke Test!
1. Add a new note
2. Type several lines with random text

- [x] Verify that pressing **Option + CMD + C** inserts a List Item wherever you click
- [x] Verify that the cursor location remains in the same relative location
- [x] Verify that typing a word + **return** results in a new List Item
- [x] Verify that pressing **tab** indents the list marker
- [x] Verify that selecting few lines of text and pressing **Option + CMD + C** turns them into List Items
- [x] Verify that pressing **CMD + Z** repeatedly revert the last OP you've performed (and the cursor's location makes sense, always!)
- [x] Verify that the Unit Tests are green 🔥 

### Release
`RELEASE-NOTES.txt` was updated in ca5150c with:
 
> Fixed (ALL) of the Editor's Undo issues when dealing with Checklists

### CODE REVIEW

File: Simplenote/NSTextView+Simplenote.swift [hunk 1]
```diff
@@ -33,6 +44,54 @@ extension NSTextView {
...
+//
+private extension NSTextView {
+
+    /// Performs an Undoable Operation `in a transactional fashion`: an Undo Group will wrap the execution of the specified Block.
+    ///
+    ///     1.  Registers an Undo Operation which is expected to restore the TextView to its previous state
+    ///     2.  Wraps up a given `Block` within an Undo Group
+    ///     3.  Post a TextDidChange Notification
+    ///
+    func performUndoableOperation(block: (NSTextStorage) -> Void) -> Bool {
```
[the remark this answers is not in the document]
Comment (diegoreymendez) [reply Simplenote/NSTextView+Simplenote.swift:57]: Really interesting related conversation: https://forum.wordreference.com/threads/undoable-something-that-can-be-undone-another-word.1683505/

File: RELEASE-NOTES.txt [hunk 2]
```diff
@@ -3,6 +3,7 @@
 - Simplenote now requires macOS 10.13 or higher
 - Updated Checklist Icons
 - Fixed a crash related to Text Linkification and Undo Support
+- Fixed (ALL) of the Editor's Undo issues when dealing with Checklists
```
[the remark this answers is not in the document]
Comment (jleandroperez) [reply RELEASE-NOTES.txt:6]: Good point!! redacting to a much humble statement!! Thank you!!!

File: Simplenote/TextViewInputHandler.swift [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    /// Lists Processing RegEx
+    ///
+    private let regexForListMarkers = NSRegularExpression.regexForListMarkers
+
+
+    /// Handles TextView's `shouldChangeTextInRange:string:` Delegate API:
+    ///
+    /// -   Note:
+    ///    This method will `only return false` (and thus, will override NSTextView's default Text Insertion) whenever the resulting
+    ///    document contains  `at least` one Markdown List Item, this method
```
[the remark this answers is not in the document]
Comment (jleandroperez) [reply Simplenote/TextViewInputHandler.swift:17]: Ahhh apologies, that's a left over from (one of many) changes. Fixing, thank you!

**Comment by jleandroperez** [comment]
Renamed `performUndoableOperation` > `registerUndoCheckpointAndPerform` and `registerUndoCheckpoint`.

Thank you both!!!

**Pull request merged**